cgrok/tree.c

```
@@ -0,0 +1,198 @@
+  if (tree->iter != NULL) {
+    dict_itor_free(tree->iter);
+  }
+  tree->iter = dict_itor_new(tree->dict);
+  dict_itor_first(tree->iter);
+}
+
+const void *tctreeiternext(TCTREE *tree, int *sp) {
+  void *key = dict_itor_key(tree->iter);
+  dict_itor_next(tree->iter);
+  *sp = 0;
+  if (key) {
+    *sp = *(uint32_t*)key;
+    return key+4;
+  }
+  return NULL;
+}
+
+void *tcpack(const void *value, uint32_t size) {
+  char *buf = malloc(size + 5);
+  memcpy(buf, &size, 4);
+  memcpy(buf+4, value, size);
+  buf[size+4] = 0;
+  return buf;
+}
+
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  void *val = tcpack(vbuf, vsiz);
+  bool inserted; 
+  void ** valPtr = dict_insert(tree->dict, key, &inserted);
+  *valPtr = val;
+}
+
+bool tctreeputkeep(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz); 
+  void *val = tcpack(vbuf, vsiz); 
+  bool inserted;
+  void **valPtr = dict_insert(tree->dict, key, &inserted);
+  if (inserted) {
+    *valPtr = val;
+  }
+  return inserted;
+}
+
+const void *tctreeget(TCTREE *tree, const void *kbuf, int ksiz, int *sp) {
+  const void *key = tcpack(kbuf, ksiz);
+  void *value = dict_search(tree->dict, key);
+  *sp = 0;
+  if (value) {
+    *sp = *(uint32_t*)value;
+    return value+4;
+  }
+  return NULL;
+}
+
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
+    dict_itor_free(tree->iter);
+  }
+  free(tree);
+}
+
+// A simple, singly linked list
+TCLIST *tclistnew(void) {
+  TCLIST *list = malloc(sizeof(TCLIST));
+  list->len = 0;
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  newElem->elem = 0;
+  list->head = newElem;
+  return list;
+}
+
+int tclistnum(const TCLIST *list) {
+  return list->len;
+}
+
+void *tclistpack(void *buf, uint32_t size) {
+  void *elem = malloc(size+1);
```

I can do that in some places (propagate the NULL up), but in methods like `tctreeput` we have no way to signal an error to the user. It seems like failing fast (with an assert like @brianc mentioned) would be better than silently continuing without having inserted the value.